Implement Imprison disabling logic using Test Fight

andrebastosdias opened the pull request:
Fixes https://github.com/smogon/pokemon-showdown/issues/6694
https://www.smogon.com/forums/threads/imprison-in-doubles.3747329/

Karthik99999 commented:
You asked a question about the past gen mechanics of this in the thread you linked. Did you end up confirming them yourself?

andrebastosdias commented:
I think this should be fine, since move requests only occur after the `DisableMove` event has been run (at the end of the turn), which may re-activate `maybeDisabled`.

Slayer95 commented:
I like the structure at 5646ef3940885cacf012a1dc70d87c85e17de7c1 better (since we aren't doing ``if (!isLastActive) this.maybeDisabled = false;`` either). That structure also allows omitting the ``!!lockedMove`` cast.

Slayer95 commented:
BTW. The bug mentioned by @DaWoblefet in that thread re: moves sometimes being greyed out from the start is because of https://github.com/smogon/pokemon-showdown/blob/master/sim/pokemon.ts#L999

This line assumes that whether a move demands a target in PS has a perfect correspondence with whether a move demands a target in cart.

andrebastosdias commented:
> BTW. The bug mentioned by @DaWoblefet in that thread re: moves sometimes being greyed out from the start is because of https://github.com/smogon/pokemon-showdown/blob/master/sim/pokemon.ts#L999
> 
> This line assumes that whether a move demands a target in PS has a perfect correspondence with whether a move demands a target in cart.

Also, I'm a bit confused about this because, in Doubles except Gen 3, every move—except for locked moves—has to go through "Select that move's target", even self-targeting moves.

Slayer95 commented:
> If that is correct, is this still desirable?

Yeah, I think so.

> Also, I'm a bit confused about this because, in Doubles except Gen 3, every move—except for locked moves—has to go through "3. Select that move's target", even self-targeting moves. So, you can just grey all Imprisoned moves for both slots. This would only be a problem in Gen 3 Doubles.

So that's how it is. I only ever playing Doubles in RSE.

andrebastosdias commented:
@Slayer95 Doesn't this fix everything?

Slayer95 commented:
Possibly :) I wish testing this were easier.

Slayer95 commented on sim/pokemon.ts:
```diff
@@ -995,16 +995,20 @@ export class Pokemon {
-				this.side.active.length >= 2 && this.battle.actions.targetTypeChoices(target!)
+				((moveSlot.pp <= 0 && !this.volatiles['partialtrappinglock']) || disabled) &&
+				(this.battle.gen >= 4 || this.battle.actions.targetTypeChoices(target!)) &&
+				this.battle.gameType !== 'singles'
```
This grouping is wrong. ``&&`` has higher priority than ``||``

andrebastosdias commented:
Give me a second. While fixing the tests, I changed it so that if you click on any move and `pokemon.maybeDisabled = true` and all moves are disabled, Struggle is automatically selected.

We would need to update the `maybeDisabled` warning in the client to inform the player that if all moves are disabled, the Pokémon will use Struggle.

Slayer95 commented:
Actually, I think the best solution for Imprison is to use the new testfight thing

- Choosing any move while ``maybeDisabled`` is still uncancellable.
- Choosing ``Test Fight``:
  - If any moves valid:
    Send message "You have some valid moves".
    If Doubles, update the request, revealing some/all disabled moves.
      - Gen 3: Reveal all disabled targeting moves.
      - Gen 4+: Reveal all disabled moves.

  - If all moves invalid:
    Force Struggle, uncancellable.

andrebastosdias commented:
Implemented @Slayer95  's solution and fixed Gen 1, which was allowing the player to cancel a move if clicked on it directly.
I still want to implement some tests and I need to fix the Dynamax test (any help would be appreciated).

I would also suggest changing the client to only show the moves if the Test Fight Button has been pressed, since it can be confusing for someone looking at it for the first time.

andrebastosdias commented:
Added `updateOnUndo` to `side#choice` so that if a player clicks on a move directly and the Pokémon is not locked, undoing that action will send an updated request without `maybeLocked`.

andrebastosdias commented:
I still want to implement some tests.
There are some existing tests I need help with: one involving Dynamax and disabled moves, and for some reason, the Exhaustive Runner is catching an error with Horn Drill having an invalid target (`targetLoc = 0`). The test never finishes.

Slayer95 commented:
Test updates are good. Changes to ``pokemon.ts`` are good. ``maybeLocked`` may need further attention.

Tested [84f2b60](https://github.com/smogon/pokemon-showdown/pull/11208/commits/84f2b60d932e8d5a4e74250dccfa039f8a1dc496)

1. All moves disabled by Imprison.
![image](https://github.com/user-attachments/assets/f811a10d-8eb6-4f92-9244-4a6e9a7998cb)
![image](https://github.com/user-attachments/assets/58617f15-e301-4633-b26d-ef8a6771b158)

Looks good, except for some client texts:
- You might be locked into a move [...] (prevents switching if you're locked) -> You might have all your moves disabled. (forces Struggle if you have no valid moves.)

2. There are valid moves disabled by imprison.
- Pressing Test Fight results in an error message: ``[Unavailable choice] Dragonite is not locked``

> Added updateOnUndo to side#choice so that if a player clicks on a move directly and the Pokémon is not locked, undoing that action will send an updated request without maybeLocked.

This means that an originally uncancellable action must be retroactively changed to undoable after the move is sent, right? That's more complex, and can potentially be more confusing. Can we centralize the request updating in the Test Fight button? Just one press of Test Fight button would be equivalent to checking all moves. That would be more efficient for players.

andrebastosdias commented:
1. All moves disabled by Imprison.
> You might be locked into a move [...] (prevents switching if you're locked) -> You might have all your moves disabled. (forces Struggle if you have no valid moves.)

That is the existing message for Gen 1. `maybeLocked` is being used for two different cases. Message should be something like `You might be locked into a move, or all your moves might be disabled and you'll be forced to struggle.` Or just send different messages depending on the generation.

3. Also, the `Cancel` button is showing up. Pressing it results in an error message, and a confused client state.

> Can we centralize the request updating in the Test Fight button? Just one press of Test Fight button would be equivalent to checking all moves. That would be more efficient for players.

Moves are all checked when you press `Test Fight`. The problem is that you can choose a move directly without using `Test Fight`, so if you cancel, you need to resend the updated request (if you can cancel, you know you’re not locked). The best solution would be to require clicking `Test Fight` in this situation.

Slayer95 commented:
We normally think of Cancel as an additional PS-exclusive convenience, that can be omitted whenever there's a risk of information leak. This problem can be avoided by just removing the ``Cancel`` ability if imprisoned, no matter whether you ultimately have or lack valid moves.

Requiring clicking ``Test Fight`` always would be too cumbersome.

andrebastosdias commented:
> We normally think of Cancel as an additional PS-exclusive convenience, that can be omitted whenever there's a risk of information leak. This problem can be avoided by just removing the Cancel ability if imprisoned, no matter whether you ultimately have or lack valid moves.

My point was, even if you have a Steel-type against a possible Magnet Pull Pokémon and you switch, the client still lets you use the cancel button and shows other options, but they will still fail if you submit them. That’s a client issue and out of scope for this PR.

> Requiring clicking Test Fight always would be too cumbersome. Ideally, players would only press it the first turn they are Imprisoned. Afterwards, they will already have the information needed to choose their moves.

That would be a bit complicated. It would be equivalent to sending `trapped` if `maybeTrapped` was confirmed as true on a previous turn. You’d have to keep track of so much stuff.

If you don’t want to click `Test Fight` every turn (which isn’t that bad since the button will be in the same place as the move buttons), you’ll have to live with the update request being sent on undo.

andrebastosdias commented:
The thing is, I separated `maybeLocked` from `maybeDisabled` since they are two different things and can exist independently. Let's recap:

* `maybeLocked` means that you may click the "Fight Button". If you are locked (including Struggle), you stay locked and can't cancel. But if you are not locked, there is no reason you shouldn't be able to cancel.
* `maybeDisabled` means that if you click on a move, you won't be able to cancel.
* `maybeTrapped` means that if you click on a switch, you won't be able to cancel.

We always need to set `maybeLocked` if `maybeDisabled` is `true`, in case all moves are disabled. But in doubles, where you can be `maybeLocked` but not `maybeDisabled` (because of the "select target" buffer), there is no reason not to allow the player to cancel the move if they pass the "Test Fight" check.

There are three solutions:

1. Send an updated request when cancelling.
2. Mandate clicking "Test Fight" (my preferred choice).
3. Make it so that `maybeLocked` does not allow cancelling (from what I understand, this is your preferred option).

Slayer95 commented:
re: 1. you intend to update requests on Cancel, but if we want to allow Cancel, we need to:
As soon as the original decision is sent, update the request telling the player that they aren't disabled, and they can Cancel (they originally couldn't). Note that this path doesn't involve updating "on undo".

re: 2. Similar to you, I intend that clicking on Test Fight would update the updated request with maybeLocked as you say. However, again, forcing 3 steps every turn just like cart is too bundersome and unneeded. No need to match clunky cart UX since information management can be covered with a better UX.

re: 3. So long as Test Fight is available, the "oppresiveness" Wob talks about in its forum post goes away. Players can still figure out which moves are legal without commiting to them, just like in cart, except that not through Undo. PS undo is very different from cart undo anyway.

andrebastosdias commented:
1. I fixed the tests.
2. I removed the updated request after a cancelled action. The "Test Fight" button is implemented the same way it is currently implemented for Gen 1. If someone wants to change how the "Test Fight" mechanics work in general for PS, they can do it in a follow-up PR. Otherwise, this will never get merged.
3. I fixed a bug where the player could Terastallize their Pokémon while using Struggle.
4. I’m waiting on research to confirm how Imprison interacts with locking moves (e.g., Outrage) and two-turn moves (e.g., Solar Beam) https://www.smogon.com/forums/threads/scarlet-violet-battle-mechanics-research.3709545/page-60#post-10613680
5. I will implement tests for the "Test Fight" button tomorrow. :)

andrebastosdias commented:
Tests added.
I changed it so that when the `Test Fight` button is clicked, instead of sending an error message, it simply sends a new request without `maybeLocked`, etc. Using the same method, I ended up re-adding the updated request when cancelling an action, but it’s a lot simpler now. See `battle#undoChoice`.

Ready for review!

Slayer95 commented on sim/side.ts:
```diff
@@ -665,9 +665,10 @@ export class Side {
+			this.updateRequestForPokemon(pokemon, req => this.updateDisabledRequest(pokemon, req));
+			this.emitRequest();
+			this.choice.error = 'Hack to avoid sending error messages to the client :D';
+			return false;
```
This hack will confuse the timer into thinking that a new turn started. https://github.com/smogon/pokemon-showdown/blob/master/server/room-battle.ts#L783-L811

andrebastosdias replied:
This also happens with trapped Pokemon
https://github.com/smogon/pokemon-showdown/blame/f4380feffe6dd7ee21f6d5adaeff4acd350d6469/sim/side.ts#L869-L881

Slayer95 commented on sim/battle.ts:
```diff
@@ -3028,6 +3028,16 @@ export class Battle {
+				}
+			}
+			if (update) side.emitRequest();
+		}
```
Same as https://github.com/smogon/pokemon-showdown/pull/11208/files/24e67db6fb1a2c2f7289ec93a530e6ec3420e336..d55a87556387b9b232e7b21b28ef47440ba22c73#r2187876948

Zarel commented:
I don't see anything obviously wrong. Thanks!

andrebastosdias commented on test/sim/moves/imprison.js:
```diff
@@ -28,13 +28,12 @@ describe('Imprison', () => {
 
 		// Imprison is not passed by Baton Pass
-		battle.makeChoices('move batonpass', 'move struggle');
+		battle.makeChoices('move batonpass', 'auto');
```
The move request includes shows all the imprisoned moves as available.
